### PR TITLE
Change flags from Int to Long, and remove hasFlag(flag: Int) method.

### PR DESCRIPTION
**Pull request by ccjoywang**

### CODE REVIEW

**Comment by ccjoywang** [comment]
moved from https://github.com/scala/scala/pull/6376

**Comment by ccjoywang** [comment]
Thanks to @rorygraves 's help comparing the performance.
We observed slightly better CPU time:
https://raw.githubusercontent.com/rorygraves/perf_tester/master/results.kept/rory/joy/output.txt
```
after 0 best 45%
                 Run Name	Cycle	samples	           Wall time (ms)	       All Wall time (ms)	                  CPU(ms)	           Idle time (ms)	           Allocated(MBs)
                     base	   1	  67	 6,059.60 [-1.56% +1.26%]	 6,059.60 [-1.56% +1.26%]	 6,019.18 [-2.20% +1.04%]	       0.00 [ NaN%  NaN%]	 1,683.11 [-0.12% +0.09%]
                     base	   2	  67	 6,070.04 [-1.68% +1.12%]	 6,070.04 [-1.68% +1.12%]	 6,021.40 [-2.20% +1.16%]	       0.00 [ NaN%  NaN%]	 1,690.94 [-0.14% +0.15%]
                     base	   3	  67	 6,061.99 [-1.90% +1.30%]	 6,061.99 [-1.90% +1.30%]	 6,015.37 [-2.50% +1.27%]	       0.00 [ NaN%  NaN%]	 1,701.38 [-0.08% +0.02%]
                     base	   4	  67	 6,026.21 [-1.25% +1.22%]	 6,026.21 [-1.25% +1.22%]	 5,980.13 [-1.81% +1.07%]	       0.00 [ NaN%  NaN%]	 1,680.41 [-0.13% +0.07%]
                     base	   5	  67	 6,200.87 [-2.00% +1.37%]	 6,200.87 [-2.00% +1.37%]	 6,156.39 [-2.67% +1.26%]	       0.00 [ NaN%  NaN%]	 1,696.87 [-0.09% +0.08%]
                    after	   1	  67	 5,954.94 [-1.56% +1.24%]	 5,954.94 [-1.56% +1.24%]	 5,906.42 [-2.11% +1.14%]	       0.00 [ NaN%  NaN%]	 1,688.62 [-0.10% +0.04%]
                    after	   2	  67	 5,877.10 [-1.41% +1.00%]	 5,877.10 [-1.41% +1.00%]	 5,833.37 [-2.01% +1.06%]	       0.00 [ NaN%  NaN%]	 1,680.84 [-0.03% +0.03%]
                    after	   3	  67	 5,975.57 [-1.37% +1.18%]	 5,975.57 [-1.37% +1.18%]	 5,925.75 [-1.85% +1.10%]	       0.00 [ NaN%  NaN%]	 1,683.89 [-0.13% +0.09%]
                    after	   4	  67	 5,981.39 [-2.45% +1.26%]	 5,981.39 [-2.45% +1.26%]	 5,933.93 [-3.00% +1.01%]	       0.00 [ NaN%  NaN%]	 1,688.69 [-0.02% +0.03%]
                    after	   5	  67	 6,130.46 [-1.56% +1.20%]	 6,130.46 [-1.56% +1.20%]	 6,081.73 [-2.07% +1.04%]	       0.00 [ NaN%  NaN%]	 1,677.15 [-0.37% +0.39%]
```

**Comment by adriaanm** [comment]
ping @retronym

**Comment by retronym** [comment]
The change looks reasonable to me. I need to run this through our benchmark suite to see if I can reproduce the gains.

Was this change driven by profiler results? How do they look before/after?

**Comment by mkeskells** [comment]
This change was an initial small task to introduce another team member into scalac contribution.
@ccjoywang has included the profilerResults that @rorygraves confirmed.

To be honest we were not expecting to see any measurable improvement from this. I never looked in detail as to the change and what actually is generated differently in the code after C2, as I would have expected that at most it is a stack frame saving (which I would expect to be inlined) and int > long widening that I would have though would be automatically done in the inlining

I expected the performance to be flat, and no worse, but happy to take the small win 😄 

We did rerun the benchmark to confirm the results a couple of times before we published

File: src/reflect/scala/reflect/internal/Flags.scala [hunk 1]
```diff
@@ -81,36 +81,36 @@ package internal
...
-  final val IMPLICIT      = 1 << 9
-  final val FINAL         = 1 << 5    // May not be overridden. Note that java final implies much more than scala final.
-  final val PRIVATE       = 1 << 2
-  final val PROTECTED     = 1 << 0
-
-  final val SEALED        = 1 << 10
-  final val OVERRIDE      = 1 << 1
-  final val CASE          = 1 << 11
-  final val ABSTRACT      = 1 << 3        // abstract class, or used in conjunction with abstract override.
+   final val IMPLICIT      = 1L << 9
```
Comment (som-snytt) [inline src/reflect/scala/reflect/internal/Flags.scala]: extra space

Comment (ccjoywang) [reply src/reflect/scala/reflect/internal/Flags.scala]: sure. will fix it.

Comment (ccjoywang) [reply src/reflect/scala/reflect/internal/Flags.scala]: sure. will fix it.

**Comment by retronym** [comment]
Sorry about the delay on this one. I suggest that we target this to 2.13.x as a few of the methods we're changing here are part of the de facto API of the compiler, and could cause unnecessary breakage in compiler plugins.

**Comment by retronym** [comment]
https://github.com/scala/scala/pull/7176

**Pull request closed without merging**